lib/libsm/sm3.c

```diff
@@ -0,0 +1,290 @@
+	GET_UINT32_BE(W[7], data, 28);
+	GET_UINT32_BE(W[8], data, 32);
+	GET_UINT32_BE(W[9], data, 36);
+	GET_UINT32_BE(W[10], data, 40);
+	GET_UINT32_BE(W[11], data, 44);
+	GET_UINT32_BE(W[12], data, 48);
+	GET_UINT32_BE(W[13], data, 52);
+	GET_UINT32_BE(W[14], data, 56);
+	GET_UINT32_BE(W[15], data, 60);
+
+#define FF0(x, y, z) ((x) ^ (y) ^ (z))
+#define FF1(x, y, z) (((x) & (y)) | ((x) & (z)) | ((y) & (z)))
+
+#define GG0(x, y, z) ((x) ^ (y) ^ (z))
+#define GG1(x, y, z) (((x) & (y)) | ((~(x)) & (z)))
+
+#define SHL(x, n)  (((x) & 0xFFFFFFFF) << (n))
+#define ROTL(x, n) (SHL((x), n) | ((x) >> (32 - n)))
+
+#define P0(x) ((x) ^ ROTL((x), 9) ^ ROTL((x), 17))
+#define P1(x) ((x) ^ ROTL((x), 15) ^ ROTL((x), 23))
+
+	for (j = 16; j < 68; j++) {
+		/*
+		 * W[j] = P1( W[j-16] ^ W[j-9] ^ ROTL(W[j-3],15)) ^
+		 *        ROTL(W[j - 13],7 ) ^ W[j-6];
+		 */
+
+		Temp1 = W[j - 16] ^ W[j - 9];
+		Temp2 = ROTL(W[j - 3], 15);
+		Temp3 = Temp1 ^ Temp2;
+		Temp4 = P1(Temp3);
+		Temp5 =  ROTL(W[j - 13], 7) ^ W[j - 6];
+		W[j] = Temp4 ^ Temp5;
+	}
+
+	for (j =  0; j < 64; j++)
+		W1[j] = W[j] ^ W[j + 4];
+
+	A = ctx->state[0];
+	B = ctx->state[1];
+	C = ctx->state[2];
+	D = ctx->state[3];
+	E = ctx->state[4];
+	F = ctx->state[5];
+	G = ctx->state[6];
+	H = ctx->state[7];
+
+	for (j = 0; j < 16; j++) {
+		SS1 = ROTL((ROTL(A, 12) + E + ROTL(T[j], j)), 7);
+		SS2 = SS1 ^ ROTL(A, 12);
+		TT1 = FF0(A, B, C) + D + SS2 + W1[j];
+		TT2 = GG0(E, F, G) + H + SS1 + W[j];
+		D = C;
+		C = ROTL(B, 9);
+		B = A;
+		A = TT1;
+		H = G;
+		G = ROTL(F, 19);
+		F = E;
+		E = P0(TT2);
+	}
+
+	for (j = 16; j < 64; j++) {
+		SS1 = ROTL((ROTL(A, 12) + E + ROTL(T[j], j)), 7);
+		SS2 = SS1 ^ ROTL(A, 12);
+		TT1 = FF1(A, B, C) + D + SS2 + W1[j];
+		TT2 = GG1(E, F, G) + H + SS1 + W[j];
+		D = C;
+		C = ROTL(B, 9);
+		B = A;
+		A = TT1;
+		H = G;
+		G = ROTL(F, 19);
+		F = E;
+		E = P0(TT2);
+	}
+
+	ctx->state[0] ^= A;
+	ctx->state[1] ^= B;
+	ctx->state[2] ^= C;
+	ctx->state[3] ^= D;
+	ctx->state[4] ^= E;
+	ctx->state[5] ^= F;
+	ctx->state[6] ^= G;
+	ctx->state[7] ^= H;
+}
+
+void sm3_update(struct sm3_context *ctx, const uint8_t *input, size_t ilen)
+{
+	size_t fill;
+	size_t left;
+
+	if (ilen <= 0)
+		return;
+
+	left = ctx->total[0] & 0x3F;
+	fill = 64 - left;
+
+	ctx->total[0] += ilen;
+	ctx->total[0] &= 0xFFFFFFFF;
+
+	if (ctx->total[0] < ilen)
+		ctx->total[1]++;
+
+	if (left && ilen >= fill) {
+		memcpy(ctx->buffer + left, input, fill);
+		sm3_process(ctx, ctx->buffer);
+		input += fill;
+		ilen -= fill;
+		left = 0;
+	}
+
+	while (ilen >= 64) {
+		sm3_process(ctx, input);
+		input += 64;
+		ilen -= 64;
+	}
+
+	if (ilen > 0)
+		memcpy(ctx->buffer + left, input, ilen);
+}
+
+static const uint8_t sm3_padding[64] = {
+	0x80, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+	0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+	0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+	0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0
+};
+
+void sm3_final(struct sm3_context *ctx, uint8_t output[32])
```

Comment:
maybe pass a pointer here and in the other sm3 API functions?

Reply:
ok

Reply:
well actually since we have a fixed size output, perhaps it's better to keep `[32]` for documentation purposes?